libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs

```diff
@@ -941,59 +942,160 @@ private List<string> SplitUtterance(string utterance, List<EntityInfo> recognize
             return unrecognized;
         }
 
-        // Combine entity values and $instance meta-data
-        private Dictionary<string, List<EntityInfo>> NormalizeEntities(ActionContext actionContext)
+        // Expand object that contains entities which can be op, property or leaf entity
+        private void ExpandEntityObject(
+            JObject entities, string op, string property, JObject rootInstance, List<string> operations, List<string> properties, uint turn, string text, Dictionary<string, List<EntityInfo>> entityToInfo)
         {
-            var entityToInfo = new Dictionary<string, List<EntityInfo>>();
-            var text = actionContext.State.GetValue<string>(TurnPath.Recognized + ".text");
-            if (actionContext.State.TryGetValue<dynamic>(TurnPath.Recognized + ".entities", out var entities))
+            foreach (var token in entities)
             {
-                var turn = actionContext.State.GetValue<uint>(DialogPath.EventCounter);
-                var operations = dialogSchema.Schema[OperationsKey]?.ToObject<List<string>>() ?? new List<string>();
-                var metaData = entities[InstanceKey];
-                foreach (var entry in entities)
+                var entityName = token.Key;
+                var instances = entities[InstanceKey][entityName] as JArray;
+                ExpandEntities(entityName, token.Value as JArray, instances, rootInstance, op, property, operations, properties, turn, text, entityToInfo);
+            }
+        }
+
+        // Expand the array of entities for a particular entity
+        private void ExpandEntities(
```

Comment:
The new entity code looks good - to the extent that I can tell without stepping through it.

However, the AdaptiveDialog class is becoming very heavy. In particular this entity processing seems like it could live in its own class, maybe "EntityHelper" or something like that. (The general internal class dependency appears to be the dialogSchema object which could easily be passed in.)

Putting the entity normalization stuff in its own separate place might also make it easier to add traditional unit tests on these functions.

Reply:
That is something we have talked about doing--in particular eventually you could imagine wanting to make the entity analysis logic replaceable, although that is a bigger change.  Do you think that splitting it out needs to be done now?

---
In reply to: [478698100](https://github.com/microsoft/botbuilder-dotnet/pull/4481#discussion_r478698100) [](ancestors = 478698100)

Reply:
Great Idea for DCR for future release. I would call it something like "EntityMapper"  I will file the DCR to add to the backlog.

---
In reply to: [478746355](https://github.com/microsoft/botbuilder-dotnet/pull/4481#discussion_r478746355) [](ancestors = 478746355,478698100)

Reply:
Filed DCR #4539 to track this.